src/native/eventpipe/ep-buffer-manager.c

```diff
@@ -1231,6 +1256,59 @@ ep_buffer_manager_write_all_buffers_to_file_v4 (
 		}
 	}
 
+	// There are sequence points created during this flush and we've marked session states for deletion.
+	// We need to remove these from the internal maps of the subsequent Sequence Points
+	if (ep_rt_thread_session_state_array_size (&session_states_to_delete) > 0) {
+		EP_SPIN_LOCK_ENTER (&buffer_manager->rt_lock, section4)
+			if (buffer_manager_try_peek_sequence_point (buffer_manager, &sequence_point)) {
+				// foreach (sequence_point in buffer_manager->sequence_point_list)
+				for (ep_rt_sequence_point_list_iterator_t sequence_point_list_iterator = ep_rt_sequence_point_list_iterator_begin (&buffer_manager->sequence_points);
+					!ep_rt_sequence_point_list_iterator_end (&buffer_manager->sequence_points, &sequence_point_list_iterator);
+					ep_rt_sequence_point_list_iterator_next (&sequence_point_list_iterator)) {
+
+					sequence_point = ep_rt_sequence_point_list_iterator_value (&sequence_point_list_iterator);
+
+					// foreach (session_state in session_states_to_delete)
+					for (ep_rt_thread_session_state_array_iterator_t thread_session_state_array_iterator = ep_rt_thread_session_state_array_iterator_begin (&session_states_to_delete);
+						!ep_rt_thread_session_state_array_iterator_end (&session_states_to_delete, &thread_session_state_array_iterator);
+						ep_rt_thread_session_state_array_iterator_next (&thread_session_state_array_iterator)) {
+
+						EventPipeThreadSessionState * thread_session_state = ep_rt_thread_session_state_array_iterator_value (&thread_session_state_array_iterator);
+						uint32_t thread_sequence_number = 0;
+						bool exists = ep_rt_thread_sequence_number_map_lookup (ep_sequence_point_get_thread_sequence_numbers_cref (sequence_point), thread_session_state, &thread_sequence_number);
+						if (exists) {
+							ep_rt_thread_sequence_number_map_remove (ep_sequence_point_get_thread_sequence_numbers_ref (sequence_point), thread_session_state);
```

Comment:
Does this not race with the call to `ep_rt_thread_sequence_number_map_free`? From what I can see, it's possible for another thread to call methods in this sequence during disable:
`disable_holding_lock` -> `ep_session_write_sequence_point_unbuffered` -> `ep_session_write_sequence_point_unbuffered` -> `ep_sequence_point_fini` -> `sequence_point_fini` -> `ep_rt_thread_sequence_number_map_free`. 

`disable_holding_lock` (or all of the subsequent callees) doesn't try to grab the buffer manager's lock and only holds the EventPipe lock, while this thread only holds the buffer manager's lock. So it looks like a use after free can happen here? Comments in SHash says:

```
// - Remove operations may be asynchronous with Lookup/Add, unless elements are also deallocated. (In which
//    case full synchronization is required)
```

Reply:
Hmmmmm.... Let me do some thinking on this. This logic _only_ happens on the streaming thread, which gets turned off before session disable. I'll look and see if I can definitively say so, though.

Reply:
I'm pretty sure we're safe here. This logic is only happening while the streaming thread is live. The disable logic disables the streaming thread before doing the unbuffered writes:

`disable_holding_lock` -> \*`ep_session_disable` (this joins the thread where the new logic is running)\* -> `ep_session_write_sequence_point_unbuffered` -> `ep_session_write_sequence_point_unbuffered` -> `ep_sequence_point_fini` -> `sequence_point_fini` -> `ep_rt_thread_sequence_number_map_free`

Reply:
As for the comment in `SHash` I think we're also good on that front. The actual dealloc happens under the thread lock, so it is only asynchronous while the object still exists.